### DISCUSSION TITLE
Why can't we use aliases within delete where clause statements

### DISCUSSION THREAD

**Discussion by frozzway** [sqlalchemy/sqlalchemy#9196] [Usage Questions] [answered]
Hello. To be brief why can we execute
```
select(U).where(U.id == 3)
```
but we can't
```
delete(U).where(U.id == 3)
```
where `U` type is `AliasedClass` and was instantiated as `U = aliased(User)` and `User` type is `DeclarativeMeta`

SQL preview of second statement:
```
DELETE FROM user AS user_1 WHERE user_1.id = :id_1
```
We would get an `sqlalchemy.orm.exc.UnmappedColumnError: No column anon_1.id is configured on mapper mapped class ...`

**Comment by zzzeek** [comment, marked as answer]
there's not a single answer here since there are several things stated that aren't really true / accurate.

firstly, I can't reproduce that error.  not known what SQLAlchemy version is in use or specifics of the model.

as for the syntax, you can DELETE against the alias, however, it does not indicate that the DELETE would proceed against that table with an alias applied, as the SQL you indicate:

    DELETE FROM user AS user_1 WHERE user_1.id = :id_1

is not standard SQL that SQLAlchemy currently does not render.  for example on MariaDB it errors out:

    MariaDB [test]> DELETE FROM a AS a_1 WHERE a_1.id = 1;
    ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near 'AS a_1 WHERE a_1.id = 1' at line 1

overall it doesnt make much sense to DELETE against an alias like that as aliases are used for stating a table multiple times in a SELECT, so current behavior is on PostgreSQL you will get this SQL:

    DELETE FROM a USING a AS a_1 WHERE a_1.id = %(id_1)s

SQLite will produce a clear error message:

    NotImplementedError: This backend does not support multiple-table criteria within DELETE

MySQL produces a statement that is effectively broken:

    DELETE FROM a USING a, a AS a_1 WHERE a_1.id = %s

this is all because SQLAlchemy, seeing that alias, assumes you want to use multiple-table deletes which are briefly introduced at https://docs.sqlalchemy.org/en/20/tutorial/data_update.html#multiple-table-deletes .

demo is below:

```py
from __future__ import annotations

from sqlalchemy import create_engine
from sqlalchemy import delete
from sqlalchemy import String
from sqlalchemy.orm import aliased
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import Session


class Base(DeclarativeBase):
    pass


class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    data: Mapped[str] = mapped_column(String(30))


#e = create_engine("sqlite://", echo=True)
e = create_engine("postgresql://scott:tiger@localhost/test", echo=True)
#e = create_engine("mysql://scott:tiger@localhost/test", echo=True)

Base.metadata.drop_all(e)
Base.metadata.create_all(e)

s = Session(e)

s.add_all([A(id=1, data='d1'), A(id=2, data='d2')])
s.commit()


aa = aliased(A)
s.execute(
    delete(aa).where(aa.id == 3).execution_options(synchronize_session=False)
)

```

**Reply by zzzeek** [reply]
I would consider the MySQL case here to be broken and we can look into fixing that.  MSSQL has the same effect for that query.   Oracle raises a clear notimplemented error like SQLite does as the syntax is not supported.

**Reply by frozzway** [reply]
As for the PostgreSQL such statement is valid and perhaps justified:
```
DELETE FROM booked_classes_table AS bk WHERE bk.date = *date* and bk.client_id IN (...)
```
So the reason I would like to use alias construct is to state the table with long name multiple times in _where_ clause statement:
```
delete(B).where(tuple_(B.attribute_name, B.another_one).in_(some_list))
```

**Reply by zzzeek** [reply]
the length of the table name is not important when using SQLAlchemy as the SQL is generated for you, it doesn't need to be typed by hand.

**Reply by zzzeek** [reply]
in any case, if you are using PostgreSQL, the statement you are doing is supported, just using the "USING" syntax which is equivalent in this case, and for MySQL / MSSQL I can fix the generation to recognize that there's only one target table.

**Reply by lelit** [reply]
> So the reason I would like to use alias construct is to state the table with long name multiple times in _where_ clause statement:
> 
> ```
> delete(B).where(tuple_(B.attribute_name, B.another_one).in_(some_list))
> ```

For that, you don't need an `alias` at the DB level, simply a shorter Python name:
```
B = very_long_and_clear_table_name
...
```

then, as Mike says, SQLAlchemy takes care of the boring length.

**Reply by frozzway** [reply]
Thanks for all the answers
>  simply a shorter Python name:

I finally realized true meaning of `alias` construct and that I don't need it. Kind of awkward of me missing its concept.
